nexus/reconfigurator/planning/src/blueprint_editor/sled_editor.rs

```diff
@@ -466,6 +492,133 @@ impl ActiveSledEditor {
+                //
+                // 1. We might not have an inventory collection for this
+                //    sled (in which case, we just skip this zone; we'll
+                //    have to backfill it during some future planning run
+                //    where we do have one)
+                // 2. sled-agent might have restarted the zone since the
+                //    inventory collection was taken and chosen a
+                //    _different_ zpool. We have no way of detecting this
+                //    now, but must be willing to overwrite a non-`None`
+                //    `filesystem_pool` value to correct ourselves if we've
+                //    hit this case (without knowing it!) in a past planner
+                //    run.
+                //
+                // We have a list of dataset names from inventory; we could
+                // try to parse those back into `DatasetName`s, but it's
+                // more straightforward to construct what this zone's
+                // `DatasetName` _would_ be (for any given zpool on the
+                // sled) and see if it's present in the inventory list. This
+                // is certainly less efficient than parsing the dataset name
+                // string, but we only have 10 zpools per sled, so shouldn't
+                // be too bad in practice.
+                let mut found_zpool = None;
+                let kind = DatasetKind::TransientZone {
+                    name: illumos_utils::zone::zone_name(
+                        zone.zone_type.kind().zone_prefix(),
+                        Some(zone.id),
+                    ),
+                };
+
+                for inv_zpool in inventory_zpools {
+                    let zpool = ZpoolName::new_external(inv_zpool.id);
+                    let dataset_name = DatasetName::new(zpool, kind.clone());
+                    let dataset_name_string = dataset_name.full_name();
+                    if inventory_datasets
+                        .iter()
+                        .any(|d| d.name == dataset_name_string)
+                    {
+                        found_zpool = Some(dataset_name.pool().clone());
+                        break;
+                    }
+                }
+
+                match found_zpool {
+                    Some(zpool) => Cow::Owned(zpool),
+                    None => {
+                        warn!(
+                            log,
+                            "could not determine expected \
+                             `filesystem_pool` for zone";
+                            "zone_id" => %zone.id,
+                            "zone_kind" => ?zone.zone_type.kind(),
+                            "current_filesystem_pool" => ?zone.filesystem_pool,
+                        );
+                        continue;
+                    }
+                }
+            };
+
+            if zone.filesystem_pool.as_ref() != Some(&*expected_filesystem_pool)
+            {
```

Comment:
Nitpick, should we use
```suggestion
            if zone.filesystem_pool.as_ref() == Some(&*expected_filesystem_pool) {
                continue;
            }
```

To avoid rightward drift?

Reply:
Yep, that's nicer: [cba6d0e](https://github.com/oxidecomputer/omicron/pull/7450/commits/cba6d0ea8f4114d61f4d15cdb419221fa2a85308)